File: core/src/main/java/com/batchsphere/core/masterdata/vendorbusinessunit/entity/VendorBusinessUnit.java
package com.batchsphere.core.masterdata.vendorbusinessunit.entity;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.math.BigDecimal;
import java.util.UUID;

@Entity
@Table(name = "vendor_business_unit")
@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class VendorBusinessUnit {

    @Id
    @Column(nullable = false, updatable = false)
    private UUID id;

    @Column(name = "vendor_id", nullable = false)
    private UUID vendorId;

    @Column(name = "unit_name", nullable = false, length = 200)
    private String unitName;

    @Column(name = "bu_code", length = 100)
    private String buCode;

    @Enumerated(EnumType.STRING)
    @Column(name = "site_type", length = 50)
    private SiteType siteType;

    @Column(columnDefinition = "TEXT")
    private String address;

    @Column(length = 100)
    private String city;

    @Column(length = 100)
    private String state;

    @Column(length = 100)
    private String country;

    @Column(length = 20)
    private String pincode;

    // ─── Site contact ──────────────────────────────────────────────────────────

    @Column(name = "site_contact_person", length = 255)
    private String siteContactPerson;

    @Column(name = "site_email", length = 255)
    private String siteEmail;

    @Column(name = "site_phone", length = 50)
    private String sitePhone;

    // ─── India regulatory ──────────────────────────────────────────────────────

    @Column(name = "drug_license_number", length = 100)
    private String drugLicenseNumber;

    @Column(name = "drug_license_expiry")
    private LocalDate drugLicenseExpiry;

    // ─── GMP Certification ─────────────────────────────────────────────────────

    @Column(name = "gmp_cert_body", length = 100)
    private String gmpCertBody;

    @Column(name = "gmp_cert_number", length = 100)
    private String gmpCertNumber;

    @Column(name = "gmp_cert_expiry")
    private LocalDate gmpCertExpiry;

    @Column(name = "is_who_gmp_certified", nullable = false)
    private Boolean isWhoGmpCertified;

    @Column(name = "is_usfda", nullable = false)
    private Boolean isUsfda;

    @Column(name = "is_eu_gmp", nullable = false)
    private Boolean isEuGmp;

    // ─── Qualification ─────────────────────────────────────────────────────────

    @Enumerated(EnumType.STRING)
    @Column(name = "qualification_status", nullable = false, length = 50)
    private QualificationStatus qualificationStatus;

    @Column(name = "qualified_date")
    private LocalDate qualifiedDate;

    @Column(name = "next_requalification_due")
    private LocalDate nextRequalificationDue;

    @Column(name = "last_audit_date")
    private LocalDate lastAuditDate;

    @Column(name = "qa_rating", precision = 3, scale = 1)
    private BigDecimal qaRating;

    @Column(name = "delivery_score", precision = 5, scale = 2)
    private BigDecimal deliveryScore;

    @Column(name = "rejection_rate", precision = 5, scale = 2)
    private BigDecimal rejectionRate;

    @Column(name = "open_capa_count")
    private Integer openCapaCount;

    @Column(name = "is_approved", nullable = false)
    private Boolean isApproved;

    @Column(name = "is_active", nullable = false)
    private Boolean isActive;

    @Column(name = "created_by", nullable = false, length = 100)
    private String createdBy;

    @Column(name = "created_at", nullable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_by", length = 100)
    private String updatedBy;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;
}

File: core/ui/src/types/vendor-business-unit.ts
export type SiteType =
  | "MANUFACTURING"
  | "TESTING_LAB"
  | "PACKAGING"
  | "DISTRIBUTION"
  | "CORPORATE_OFFICE";

export const SITE_TYPE_LABELS: Record<SiteType, string> = {
  MANUFACTURING:    "Manufacturing",
  TESTING_LAB:      "Testing Lab",
  PACKAGING:        "Packaging",
  DISTRIBUTION:     "Distribution",
  CORPORATE_OFFICE: "Corporate Office",
};

export type QualificationStatus =
  | "NOT_STARTED"
  | "APPLICATION_SUBMITTED"
  | "DOCUMENT_REVIEW"
  | "AUDIT_SCHEDULED"
  | "AUDIT_IN_PROGRESS"
  | "CAPA_PENDING"
  | "QUALIFIED"
  | "RE_QUALIFICATION_DUE"
  | "SUSPENDED"
  | "DISQUALIFIED";

export const QUAL_STATUS_LABELS: Record<QualificationStatus, string> = {
  NOT_STARTED:           "Not Started",
  APPLICATION_SUBMITTED: "Application Submitted",
  DOCUMENT_REVIEW:       "Doc Review",
  AUDIT_SCHEDULED:       "Audit Scheduled",
  AUDIT_IN_PROGRESS:     "Audit In Progress",
  CAPA_PENDING:          "CAPA Pending",
  QUALIFIED:             "Qualified",
  RE_QUALIFICATION_DUE:  "Re-qual Due",
  SUSPENDED:             "Suspended",
  DISQUALIFIED:          "Disqualified",
};

export type VendorDocumentType =
  | "COMPANY_REGISTRATION"
  | "GST_CERTIFICATE"
  | "PAN_CARD"
  | "ISO_CERTIFICATE"
  | "GMP_CERTIFICATE"
  | "DRUG_MANUFACTURING_LICENSE"
  | "COA_TEMPLATE"
  | "MSDS"
  | "VENDOR_QUESTIONNAIRE"
  | "SITE_MASTER_FILE"
  | "QUALITY_AGREEMENT"
  | "NDA"
  | "AVL_AGREEMENT"
  | "AUDIT_REPORT";

export const VENDOR_DOCUMENT_TYPE_LABELS: Record<VendorDocumentType, string> = {
  COMPANY_REGISTRATION: "Company Registration",
  GST_CERTIFICATE: "GST Certificate",
  PAN_CARD: "PAN Card",
  ISO_CERTIFICATE: "ISO Certificate",
  GMP_CERTIFICATE: "GMP Certificate",
  DRUG_MANUFACTURING_LICENSE: "Drug Manufacturing License",
  COA_TEMPLATE: "CoA Template",
  MSDS: "MSDS",
  VENDOR_QUESTIONNAIRE: "Vendor Questionnaire",
  SITE_MASTER_FILE: "Site Master File",
  QUALITY_AGREEMENT: "Quality Agreement",
  NDA: "NDA",
  AVL_AGREEMENT: "AVL Agreement",
  AUDIT_REPORT: "Audit Report",
};

export type VendorDocumentStatus = "VALID" | "EXPIRING_SOON" | "EXPIRED";

export const VENDOR_DOCUMENT_STATUS_LABELS: Record<VendorDocumentStatus, string> = {
  VALID: "Valid",
  EXPIRING_SOON: "Expiring Soon",
  EXPIRED: "Expired",
};

export type VendorAuditType =
  | "INITIAL_QUALIFICATION"
  | "PERIODIC"
  | "FOR_CAUSE"
  | "POST_CAPA"
  | "REMOTE";

export const VENDOR_AUDIT_TYPE_LABELS: Record<VendorAuditType, string> = {
  INITIAL_QUALIFICATION: "Initial Qualification",
  PERIODIC: "Periodic",
  FOR_CAUSE: "For Cause",
  POST_CAPA: "Post CAPA",
  REMOTE: "Remote",
};

export type VendorAuditStatus = "SCHEDULED" | "IN_PROGRESS" | "COMPLETED" | "CANCELLED";

export const VENDOR_AUDIT_STATUS_LABELS: Record<VendorAuditStatus, string> = {
  SCHEDULED: "Scheduled",
  IN_PROGRESS: "In Progress",
  COMPLETED: "Completed",
  CANCELLED: "Cancelled",
};

export type VendorAuditOutcome =
  | "APPROVED"
  | "APPROVED_WITH_OBSERVATIONS"
  | "REJECTED"
  | "PENDING_CAPA";

export const VENDOR_AUDIT_OUTCOME_LABELS: Record<VendorAuditOutcome, string> = {
  APPROVED: "Approved",
  APPROVED_WITH_OBSERVATIONS: "Approved w/ Observations",
  REJECTED: "Rejected",
  PENDING_CAPA: "Pending CAPA",
};

export type VendorBusinessUnit = {
  id: string;
  vendorId: string;
  unitName: string;
  buCode: string | null;
  siteType: SiteType | null;
  address: string | null;
  city: string | null;
  state: string | null;
  country: string | null;
  pincode: string | null;
  siteContactPerson: string | null;
  siteEmail: string | null;
  sitePhone: string | null;
  drugLicenseNumber: string | null;
  drugLicenseExpiry: string | null;
  gmpCertBody: string | null;
  gmpCertNumber: string | null;
  gmpCertExpiry: string | null;
  isWhoGmpCertified: boolean;
  isUsfda: boolean;
  isEuGmp: boolean;
  qualificationStatus: QualificationStatus;
  qualifiedDate: string | null;
  nextRequalificationDue: string | null;
  lastAuditDate: string | null;
  qaRating: number | null;
  deliveryScore: number | null;
  rejectionRate: number | null;
  openCapaCount: number | null;
  isApproved: boolean;
  isActive: boolean;
  createdBy: string;
  createdAt: string;
  updatedBy: string | null;
  updatedAt: string | null;
};

export type CreateVendorBusinessUnitRequest = {
  unitName: string;
  buCode?: string;
  siteType?: SiteType;
  address?: string;
  city?: string;
  state?: string;
  country?: string;
  pincode?: string;
  siteContactPerson?: string;
  siteEmail?: string;
  sitePhone?: string;
  drugLicenseNumber?: string;
  drugLicenseExpiry?: string;
  gmpCertBody?: string;
  gmpCertNumber?: string;
  gmpCertExpiry?: string;
  isWhoGmpCertified?: boolean;
  isUsfda?: boolean;
  isEuGmp?: boolean;
  qualifiedDate?: string;
  nextRequalificationDue?: string;
  lastAuditDate?: string;
  qaRating?: number;
  deliveryScore?: number;
  rejectionRate?: number;
  openCapaCount?: number;
  createdBy: string;
};

export type UpdateVendorBusinessUnitRequest = {
  unitName: string;
  buCode?: string;
  siteType?: SiteType;
  address?: string;
  city?: string;
  state?: string;
  country?: string;
  pincode?: string;
  siteContactPerson?: string;
  siteEmail?: string;
  sitePhone?: string;
  drugLicenseNumber?: string;
  drugLicenseExpiry?: string;
  gmpCertBody?: string;
  gmpCertNumber?: string;
  gmpCertExpiry?: string;
  isWhoGmpCertified?: boolean;
  isUsfda?: boolean;
  isEuGmp?: boolean;
  qualificationStatus?: QualificationStatus;
  qualifiedDate?: string;
  nextRequalificationDue?: string;
  lastAuditDate?: string;
  qaRating?: number;
  deliveryScore?: number;
  rejectionRate?: number;
  openCapaCount?: number;
  updatedBy: string;
};

export type VendorBusinessUnitDocument = {
  id: string;
  buId: string;
  documentType: VendorDocumentType;
  documentTitle: string;
  fileName: string;
  storagePath: string;
  uploadedAt: string;
  expiryDate: string | null;
  status: VendorDocumentStatus;
  uploadedBy: string;
};

export type CreateVendorBusinessUnitAuditRequest = {
  auditType: VendorAuditType;
  scheduledDate: string;
  completedDate?: string;
  auditedBy: string;
  status?: VendorAuditStatus;
  outcome?: VendorAuditOutcome;
  observationCount?: number;
  criticalObservationCount?: number;
  notes?: string;
};

export type VendorBusinessUnitAudit = {
  id: string;
  buId: string;
  auditType: VendorAuditType;
  scheduledDate: string;
  completedDate: string | null;
  auditedBy: string;
  status: VendorAuditStatus;
  outcome: VendorAuditOutcome | null;
  observationCount: number | null;
  criticalObservationCount: number | null;
  notes: string | null;
  createdAt: string;
  updatedBy: string | null;
  updatedAt: string | null;
};
